libzfs_core/_libzfs_core.py

```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+    return result
+
+
+@_uncommitted(lzc_list)
+def lzc_list_children(name):
+    '''
+    List the children of the ZFS dataset.
+
+    :param str name: the name of the dataset.
+    :return: an iterator that produces the names of the children.
+    '''
+    (fd, other_fd) = lzc_list(name, {'recurse': 1, 'type': {'filesystem': None, 'volume': None}})
+
+    def _iterator():
+        try:
+            while True:
+                record_bytes = os.read(fd, _PIPE_RECORD_SIZE)
+                if not record_bytes:
+                    break
+                (size, _, err, _, _) =  struct.unpack(_PIPE_RECORD_FORMAT, record_bytes)
+                xlate.lzc_list_children_xlate_error(err, name)
+                if size == 0:
+                    break
+                data_bytes = os.read(fd, size)
+                result = {}
+                with nvlist_out(result) as nvp:
+                    ret = _lib.nvlist_unpack(data_bytes, size, nvp, 0)
+                if ret != 0:
+                    raise ZFSError(ret, "Failed to unpack list data")
+                child = result['name']
+                if child != name:
+                    yield child
+        finally:
+            os.close(other_fd)
+            os.close(fd)
+
+    return _iterator()
+
+
+@_uncommitted(lzc_list)
+def lzc_list_snaps(name):
+    '''
+    List the snapshots of the ZFS dataset.
+
+    :param str name: the name of the dataset.
+    :return: an iterator that produces the names of the snapshots.
+    '''
+    (fd, other_fd) = lzc_list(name, {'recurse': 1, 'type': {'snapshot': None}})
+
+    def _iterator():
+        try:
+            while True:
+                record_bytes = os.read(fd, _PIPE_RECORD_SIZE)
+                if not record_bytes:
+                    break
+                (size, _, err, _, _) =  struct.unpack(_PIPE_RECORD_FORMAT, record_bytes)
+                xlate.lzc_list_snaps_xlate_error(err, name)
+                if size == 0:
+                    break
+                data_bytes = os.read(fd, size)
+                result = {}
+                with nvlist_out(result) as nvp:
+                    ret = _lib.nvlist_unpack(data_bytes, size, nvp, 0)
+                if ret != 0:
+                    raise ZFSError(ret, "Failed to unpack list data")
+                snap = result['name']
+                if snap != name:
+                    yield snap
+        finally:
+            os.close(other_fd)
+            os.close(fd)
+
+    return _iterator()
+
+
+def is_supported(func):
+    '''
+    Check whether C *libzfs_core* provides implementation required
+    for the given Python wrapper.
+
+    :param function func: the function to check.
+    :return bool: whether the funciton can be used.
+
+    If `is_supported` returns ``False`` for the function, then
+    calling the function would result in :exc:`NotImplementedError`.
+    '''
+    try:
+        # woulf fail if is_supported is not set yet
+        return func._is_supported
```

Comment:
I'm unsure the optimization of caching this result on the function object is worthwhile.  I suspect there won't be a lot of inner loops re-checking whether some function is supported.  And since the result won't ever change they should probably just hoist the check if they care about being faster (particularly since function calls in Python are often more expensive than the logic inside the function).

`is_supported` would be simpler and the `_is_supported` cache attribute would be avoided without the caching behavior, so I'd probably drop it.